core/src/avm2/globals/flash/utils.rs

```diff
@@ -140,6 +140,56 @@ pub fn escape_multi_byte<'gc>(
     Ok(v.into())
 }
 
+fn handle_percent<I>(it: &mut I) -> Option<u8>
+where
+    I: Iterator<Item = char>,
+{
+    let mut d = it.next()?.to_digit(16)?;
+    d = d * 16 + it.next()?.to_digit(16)?;
+    Some(d as u8)
+}
+
+/// Implements `flash.utils.unescapeMultiByte`
+pub fn unescape_multi_byte<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    _this: Option<Object<'gc>>,
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error> {
+    let s = args
+        .get(0)
+        .unwrap_or(&Value::Undefined)
+        .coerce_to_string(activation)?;
+    let bs = s.as_wstr();
+    let mut buf = WString::new();
+    let mut it = bs.chars().map(|c| c.unwrap_or(char::REPLACEMENT_CHARACTER));
+
+    let mut bytes = Vec::new();
```

Comment:
Instead of allocating a new buffer everytime we encounter a `%`, is there a way we can write the bytes directly into the WString without this?

Reply:
The main problem is that the input string can mix percent-escaped bytes with unescaped chars, including wide ones. So what we write to the string must be already normalized to wide chars.

These runs of bytes need to be converted to utf-16. It is possible to avoid the allocation and have a fixed buffer (like 4 bytes), then accumulate the bytes until either:

* they form a valid utf-8 encoding of a char
* they form an invalid utf-8 sequence
* an unescaped char comes next
* the input string ends

Basically create a version of`DecodeAvmUtf8` which accepts bytes one by one.

My opinion: only makes sense if this mechanism (i mean "decode utf-8 as you go") is usable elsewhere, otherwise probably not worth it.

Reply:
I think it's good enough to keep as-is. But for code clarity, I think `bytes` can be defined in a narrower scope, if we'd have peekable chars iterator. Something like:
```rs
// ...
if c == '%' {
    let mut bytes = Vec::new();
    while let Some(b) = handle_percent() {
        bytes.push(b);
        if !matches!(chars.peek(), Some('%')) {
            break;
        }
    }
    buf.push_str(&WString::from_utf8_bytes(bytes));
}
// ...
```

Reply:
Done.